packages/caver-contract/src/index.d.ts

```diff
@@ -0,0 +1,139 @@
+
+    The caver-js library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with the caver-js. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { Wallet } from '../../..'
+import { Result } from '../../caver-abi/src'
+import { KIP37 } from '../../caver-kct/src'
+import KIP7 from '../../caver-kct/src/kip7'
+import { AbiItem } from '../../caver-utils/src'
+import KeyringContainer from '../../caver-wallet/src'
+
+export interface ContractOptions {
+    address: string
+    jsonInterface: AbiItem[]
+    from: string
+    feePayer: string
+    feeDelegation: boolean
+    feeRatio: number
+    gasPrice: number
+    gas: number
+    data: string
+}
+
+export interface DeployOptions extends SendOptions {
+    data?: string
+    arguments?: any[]
+}
+
+export interface SendOptions {
+    from?: string
+    feeDelegation?: boolean
+    feePayer?: string
+    feeRatio?: number
+    gas?: number | string
+    contractDeployFormatter?: Function
+}
+
+export interface CallOptions {
+    from?: string
+    gasPrice?: string
+    gas?: number
+}
+
+export interface ContractSendMethod extends SendData {
+    send: Contract['send']
+    call: Contract['call']
+    sign: Contract['sign']
+    signAsFeePayer: Contract['signAsFeePayer']
+}
+
+export interface SendData {
+    from?: string
+    status?: string
+    type?: string
+    feePayer?: string
+    contractAddress?: string
+    feeRatio?: number
+    options?: {
+        address: string
+    }
+}
+
+export interface EventOptions {
+    filter?: object
+}
+
+export interface EventData {
+    address?: string
+    blockNumber?: number
+    transactionHash?: string
+    transactionIndex?: number
+    blockHash?: string
+    logIndex?: number
+    id?: string
+    returnValues?: {
+        operator?: string
+        from?: string
+        to?: string
+        id?: string
+        ids?: string[]
+        value?: string
+        values?: string[]
+        account?: string
+        approved?: boolean
+        tokenId?: string
+    }
+    event?: string
+    signature?: string
+    raw?: {
+        data?: string
+        topics?: string[]
+    }
+}
+
+export default class Contract {
+    constructor(jsonInterface: AbiItem[], address?: string, options?: ContractOptions)
+
+    options: ContractOptions
+    defaultSendOptions: ContractOptions
+    defaultAccount: string | null
+    defaultBlock: string
+    events: any
+    methods: any
+    _wallet: Wallet
+
+    setKeyrings(keyrings: KeyringContainer): void
+    setWallet(wallet: Wallet): void
+    addAccounts(accounts: string[]): void
+    decodeFunctionCall(functionCall: string): Result
+    deploy(options: DeployOptions, byteCode?: string, ...args: any[]): ContractSendMethod
```

Comment:
```suggestion
    deploy(options: SendOptions, byteCode: string, ...args: any[]): Promise<Contract>
    deploy(options: DeployOptions): ContractSendMethod
```

Reply:
Oh, i think this is a bug, i will fix it.

Reply:
Oh, sorry this is not a bug;.

In fact, in the code used in the test code, the last parameter is passed as `contractDeployFormatter` which will replace the default `contractDeployFormatter` function after deployement.
Depending on how `contractDeployFormatter` function in the `SendOptions` is defined, the return value can be various type.

Then maybe can we edit like below?
`deploy(options: DeployOptions, byteCode: string, ...args: any[]): Promise<any>`

Reply:
That throws an error where `caver.kct.kip37.js` 1039 line.
Error message is `Property 'send' does not exist on type 'Promise<any>'.ts(2339)`.
If use chaining methods it need clearly types or just `any`.

And `byteCode` is not a required property according to lines 1035-1038 of the `caver.kct.kip37.js` test.

Therefore, How about below?

`deploy(options: ContractDeployParams, byteCode?: string, ...args: any[]): any`
(DeployOptions changed to ContractDeployParams as your suggestion)

Reply:
If so, then the first parameter has to accept `SendOptions`. 
Maybe like this? `deploy(options: ContractDeployParams | SendOptions, byteCode?: string, ...args: any[]): any`.

Reply:
`ContractDeployParams` is extended from `SendOptions`.
Is there any reason to take the two parameters separately?

Reply:
Actually i don't understand why `ContractDeployParams` extends `SendOptions`.
Are additional fields defined in ContractDeployParams used in conjunction with fields defined in SendOptions?

Reply:
I wrote that based on `caver.contract.js` test files.
https://docs.klaytn.com/bapp/sdk/caver-js/v1.4.1/api-references/caver.klay.contract#deploy says only `data` and `arguments` but test isn't
For example, 1068-1072 getting `SendOptions` as parameter. and 1090-1093 getting `data` and `arguments` as parameter.
So, i thought `ContractDeployParams` extends `SendOptions`.
There was no place that said what properties `SendOptions` had, so I could only infer.

I'm not sure because I can't see the server.
If I'm wrong please correct me. 😅

Reply:
And, i suggest below. return `any` type is worst in typescript i think. 😓
I guess, value that added as an option is will be return. I can't test it. because `process.env.privateKey` for test wasn't served. so I'm just assuming.

```
export interface DeployedTransactionObject {
    arguments?: any[]
    send?: Contract['send']
    sign?: Contract['sign']
    signAsFeePayer?: Contract['signAsFeePayer']
    estimateGas?: RpcCallToMethod['klay_estimateGas']
}

export interface DeployedWithOptionsTransactionObject extends DeployedTransactionObject, SendData { }

deploy(options: ContractDeployParams, byteCode?: string, ...args: any[]): DeployedTransactionObject
deploy(options: SendOptions, byteCode?: string, ...args: any[]): DeployedWithOptionsTransactionObject
```

Reply:
As you said, deploy returns `DeployedTransactionObject` when [ContractDeployParams is entered as a parameter](https://docs.klaytn.com/bapp/sdk/caver-js/api-references/caver.contract#mycontract-deploy), that is, for the following usability. In this case, byteCode and args are not accepted as parameters.
`contract.deploy({ data, arguments }).send({ from, gas, ... })`

However, when [using SendOptions as a first parameter](https://docs.klaytn.com/bapp/sdk/caver-js/api-references/caver.contract#mycontract-deploy2), byteCode is a required value, and the returned value is basically `DeployedWithOptionsTransactionObject`, but the return value may vary depending on the `contractDeployFormatter` defined in SendOptions.
`contract.deploy({ from, gas, ... }, byteCode, ...args)`

```
deploy(options: ContractDeployParams): ContractMethod
deploy(options: SendOptions, byteCode: string, ...args: any[]): any
```

And actually i think that `ContractMethod` does not need to extends `SendData`.
Also i feel like `SendData` is combination of TransactionReceipt and Contract, so i hope you can use another object which defines TransactionReceipt.

To run the test, you can create `.env` file in caver-js working directory, and add below contents and fill the form.

```
EN_RPC_URL_DEV=''
EN_WS_URL_DEV=''

EN_RPC_URL=''
EN_WS_URL=''

privateKey=''
privateKey2=''
```

Reply:
First, when I tried with my key, I got the following error message.
`Error: Returned error: unknown account`
This key is valid. I can login [Klaystagram](https://klaystagram.vercel.app/) with this key.

Second, deploy with sendOptions will return `this.methods.constructor(byteCode, ...params).send(sendOptions)`
It means return execution value of `send` that `Promise<TransactionReceipt>`
So, what about below?(`ContractMethod` is removed)

```
export interface DeployedTransactionObject {
    arguments?: any[]
    send?: Contract['send']
    sign?: Contract['sign']
    signAsFeePayer?: Contract['signAsFeePayer']
    estimateGas?: RpcCallToMethod['klay_estimateGas']
    encodeABI?: () => string
}

deploy(options: ContractDeployParams): DeployedTransactionObject
deploy(options: SendOptions, byteCode: string, ...args: any[]): Promise<TransactionReceipt>
```

Reply:
No. I keep saying that if you use SendOptions as a parameter, the default returned value is a Contract instance.

In the case of the test case, because a "function" that returns a `TransactionReceipt` instead of a Contract instance is defined and passed to the `contractDeployFormatter` field inside `SendOptions` for testing, instead of a `Contract instance`, `TransactionReceipt` is returned.

By default deploy with SendOptions returns a Contract instance.

However, as mentioned above, the returned value may be different depending on the function defined in the `contractDeployFormatter` field.
This allows users to define and use functions according to the information they want to obtain after deploying the contract.
Therefore, some users can only take the address of the contract, other users can take the contract instance containing the deployed contract information as it is, and other users can receive a TransactionReceipt containing the distributed transaction result.

Reply:
And also there is `ContractMethod` still. Will you remove that?
What i think is in `ContractMethod`, some fields should be defined addtionally, and deploy with `ContractDeployParams` should return `ContractMethod`. not separate `DeployedTransactionObject`.

Reply:
Yes, I want to chagne `ContractMethod` to `DeployedTransactionObject` that based on https://docs.klaytn.com/bapp/sdk/caver-js/api-references/caver.contract#mycontract-deploy
(Document said it return Promise, but i think it's wrong, not Promise but just object)

What I wrote`ContractMethod` is wrong. it shouldn't include `call` method and it should includes `arguments`(it's not method but property), `estimateGas`, `encodeABI`
If the interface name is not appropriate, please suggest.

In short,
`deploy(options: ContractDeployParams): DeployedTransactionObject // OK`
And, with `SendOptions` it return basically `Contract` instance. But if you define `contractDeployFormatter`, it will be returned dynamically types according to `contractDeployFormatter`. right?

If what i said is right, it return `Contract` or `TransactionReceipt`(all things in the `TransactionReceipt`, refined according to the `contractDeployFormatter`)

So, what about below?
```
deploy(options: ContractDeployParams): DeployedTransactionObject
deploy(options: SendOptions, byteCode: string, ...args: any[]): Promise<Contract> | Promise<TransactionReceipt>
```

Of course, the return value may different depending on the formatter, but I think it is better to give all options than no options.

Reply:
Oh, right. That won't return Promise. i will fix it.

umm... if so, then i think we need two classes. the one is `ContractMethod` which includes the field for return object of [contract.deploy](https://docs.klaytn.com/bapp/sdk/caver-js/api-references/caver.contract#mycontract-deploy) and [contract.methods](https://docs.klaytn.com/bapp/sdk/caver-js/api-references/caver.contract#mycontract-methods).
I think name can be `ContractMethod` and `ContractMethodToDeploy`. If there is better idea, please let me know.

And yes, exactly what you explain is correct.
Um!! if we define like that, can user use `contractDeployFormatter` to receive other type return? I feel like typescript will block user to use like that.

Reply:
There seems to be some confusion in communication due to many changes, so I committed the modified file first.
please check 49c0c69.

Can you give me another example using formatter?
The test file is set up as below
```
const contractDeployFormatter = receipt => {
    return receipt
}
```
Therefore, I set return type to `Promise<TransactionReceipt>`
If it can use like below?
```
const contractDeployFormatter = receipt => {
    return receipt.status // this should return type boolean
}
```
This return type to `Promise<boolean>`
if this is right, then we should set return type to `Promise<any>` like belows
```
deploy(options: ContractDeployParams): DeployedTransactionObject
deploy(options: SendOptionsWithFormatter, byteCode: string, ...args: any[]): Promise<any>
deploy(options: SendOptions, byteCode: string, ...args: any[]): Promise<Contract>
```

Reply:
There seems to be some confusion in communication due to many changes, so I committed the modified file first.
please check 49c0c69.

First, `Contract` has `methods` property that type is `any`.
(because it depends on the jsonInterface input to the contract. and I'm not good at understanding of this.
If there is a good way, please suggest.)

Second, Can you give me another example using formatter?
The test file is set up as below
```
const contractDeployFormatter = receipt => {
    return receipt
}
```
Therefore, I set return type to `Promise<TransactionReceipt>`
If it can use like below?
```
const contractDeployFormatter = receipt => {
    return receipt.status // this should return type string
}
```
Is this return type to `Promise<string>` ?
then we should set return type to `Promise<any>` as you said, it can't predict type.
(unable to test because of the previously mentioned error)
```
deploy(options: ContractDeployParams): DeployedTransactionObject
deploy(options: SendOptionsWithFormatter, byteCode: string, ...args: any[]): Promise<any>
deploy(options: SendOptions, byteCode: string, ...args: any[]): Promise<Contract>
```

But, If the property returned in the form of an object like below
```
{
    status: 'xxx'
}
```
If then, it is as part of the `TransactionReceipt`. therefore, below is better
```
deploy(options: ContractDeployParams): DeployedTransactionObject
deploy(options: SendOptionsWithFormatter, byteCode: string, ...args: any[]): Promise<TransactionReceipt>
deploy(options: SendOptions, byteCode: string, ...args: any[]): Promise<Contract>
```

And, are you want to change `DeployedTransactionObject` to `ContractMethod`?
I think, it doesn't matter, except that `arguments` is property, not method.
Also, since it contains multiple methods `ContractMethods` is better I think.

Reply:
There seems to be some confusion in communication due to many changes, so I committed the modified file first.

Why `ContractMethods` have to includes contract.methods?
`Contract` has `methods` property that type is `any`.
(because it depends on the jsonInterface input to the contract. and I'm not good at understanding of this.
If there is a good way, please suggest.)

Since the type cannot be inferred when the`contractDeployFormatter` is used, it is fixed to handle it by overloading.
please check e2e5e45, df7cb20

Reply:
Um.. i think i make you confuse. 
First, `contract.deploy` will return `ContractMethod`. Please rename `ContractMethods` to `ContractMethod`.

And `contract.methods` is not `ContractMethod`. Actually, `contract.methods` are function which will return `ContractMethod` which can be executed.
For example, `contract.method.set('key', 'value').send({from, gas})`, `contract.method.set('key', 'value')` will return `ContractMethod` and user can send transaction via calling the function in the ContractMethod (`.send({from, gas})`).

So methods can be Map<string, (...*) => ContractMethod>.
But i dont know how this will affect to the actual code, so `any` type will be fine also.

Reply:
If so, it can be below.
```
export default class Contract {
    ...
    methods: {
        [methodName: string]: (...args: any) => ContractMethod
    }
    ....
```

But, `ContractMethod` doesn't have `call` function.
It will throw error the test CAVERJS-UNIT-ETC-275, CAVERJS-UNIT-ETC-276, etc..
To avoid this problem, need to add `call` function to `ContractMethod` or replace it with another interface.

And overloading for `contractDeployFormatter` is okay?

Reply:
Yes, i think `ContractMethod` should have `call` inside. maybe can be optional.

And sorry for making you confused. for `sign` functions in `Contract`, they will return `Transaction` which includes sender's `signatures` in the object. This will be same with `signAsFeePayer`.
They will not handling anything with `contractDeployFormatter`. 
They will always return `Transaction`, so you can keep just second definition, and just fix the return type.

`KeyringContainer` will operate with `Transaction`.
`TransactionReceipt` will be happend when transaction is sent to the Klaytn network.
So, if some functions are operated via `klay` rpc call, then high possibility that function will return `TransactionReceipt`.
If some functions will not send transaction to the Klaytn network and just still handled by caver-js, then `Transaction` will be used.
`KeyringContainer` is in-memroy wallet of caver-js, so this will use `Transaction`.
And for `Contract`, if a function will send transaction to the Klaytn network (ex. `send`, `deploy`), then this will use `TransactionReceipt` because user want to send transaction and then they want to know the result.
And for `Contract`, if some functions won't send transaction to the Klaytn (ex. `sign`, `signAsFeePayer`), then those will use `Transaction` because still handled in the caver-js.

You don't have to feel sorry for this! 
I agree that the package is massive and confusing.
I implemented this, but I'm also confused lol...
We always appreciate your contribution.
Sorry for making you confused;ㅠ

Reply:
- Can I use `call` with no `functionName` parameter? doc saying it's not optional, but test 'CAVERJS-UNIT-ETC-275', 'CAVERJS-UNIT-ETC-276'... are using without it.
(https://docs.klaytn.com/bapp/sdk/caver-js/api-references/caver.contract#mycontract-call)

- `klay_sendRawTransaction` is receving `string | object` as parameter. what is object? currently, `klay_sendRawTransaction` is written as follow
`klay_sendRawTransaction(rawTransaction: string | ValueTransfer): Promise<TransactionReceipt>`
(https://docs.klaytn.com/bapp/sdk/caver-js/api-references/caver.rpc/klay#caver-rpc-klay-sendrawtransaction)

Reply:
Sorry for late replying. ;-;..

- For `CAVERJS-UNIT-ETC-275` and `CAVERJS-UNIT-ETC-276`, `call` function can be called without `functionName`, because that `call` function is the method of the `ContractMethod`, not `Contract`.
So, if you see the test code, you can see `contract.methods.get(keyString).call({ from: sender.address })` and `contract.methods['get'](keyString).call({ from: sender.address })`, which means they will call a function named `get`.

- According to my intention, Transaction is a type that refers to all transaction types used in caver-js.
So, Transaction is a type that includes all of LegacyTransaction, ValueTransfer, ValueTransferMemo, AccountUpdate, ... FeeDelegatedChainDataAnchoringWithRatio.
AbstractTransaction is a little more concrete and is an abstract class defined for transactions.
All transactions defined in caver-js (provided in caver-transaction package) are implemented by extending AbstractTransaction.
So you can represent any transaction type by using `AbstractTransaction` when representing the type.
If I give a strange answer because it's been too long since I've confused the context, please tell me.

- Yes that's right!!! `klay_sendRawTransaction` may receive an RLP-encoded transaction string, but it may also accept transactions provided by the caver-transaction package.

-`AbstractTransaction` does not define `signAsFeePayer` as an abstract method because it also includes transactions that are not fee payment.
Instead, fee payment transaction is implemented by extending `AbstractFeeDelegatedTransaction`, and this `AbstractFeeDelegatedTransaction` is defined by extends `AbstractTransaction`.
Therefore, `signAsFeePayer` of `ContractMethod` returns a fee payment transaction type, which can be defined as `AbstractFeeDelegatedTransaction`.
And transactions ending with `~WithRatio` are transactions that can determine the ratio of fee payment.
These are implemented by extending the abstract class called `AbstractFeeDelegatedWithRatioTransaction`, but `AbstractFeeDelegatedWithRatioTransaction` is implemented by extending `AbstractFeeDelegatedTransaction`, so `AbstractFeeDelegatedTransaction` also includes partial fee delegated transaction types.

I recently added content for jsdoc.
If you rebase this PR to the latest dev and see the contents defined there, I think it will be very helpful in defining the type.